diffrax/integrate.py

```diff
@@ -142,34 +218,107 @@ def body_fun(state, inplace):
+                    _, _, _, _, _, pred, step = _val
+                    return (implicit_step & pred) | (
+                        jnp.invert(implicit_step) & (step == 0)
+                    )
 
-            def _apply_history(x):
-                if is_vf_wrapper(x):
-                    vector_field = jtu.Partial(x.vector_field, history=history_vals)
-                    return VectorFieldWrapper(vector_field)
-                else:
-                    return x
+                def _body_fun(_val):
+                    y_prev, _, dense_info, _, _, _, step = _val
+                    is_vf_wrapper = lambda x: isinstance(x, VectorFieldWrapper)
+
+                    def _apply_history(x):
+                        if is_vf_wrapper(x):
+                            vector_field = _HistoryVectorField(
+                                x.vector_field,
+                                t0,
+                                state.tprev,
+                                state.tnext,
+                                dense_info,
+                                dense_interp,
+                                solver.interpolation_cls,
+                                y0_history,
+                                delays.delays,
+                            )
+                            return VectorFieldWrapper(vector_field)
+                        else:
+                            return x
+
+                    terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
+                    (y, y_error, dense_info, solver_state, solver_result) = solver.step(
+                        terms_,
+                        state.tprev,
+                        state.tnext,
+                        state.y,
+                        args,
+                        state.solver_state,
+                        state.made_jump,
+                    )
 
-            terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
-            # TODO: write down implicit problem wrt dense_info, using `terms_`
-            (y, y_error, dense_info, solver_state, solver_result) = terms_  # ...
+                    _pred = (((y - y_prev) / y) > delays.eps).any()
+                    _pred = _pred & (step < 10)
+                    return (
+                        y,
+                        y_error,
+                        dense_info,
+                        solver_state,
+                        solver_result,
+                        _pred,
+                        step + 1,
+                    )
+
+                _init_val = (
+                    state.y,
+                    0,
+                    dict(
+                        zip(
+                            state.dense_infos.keys(),
+                            [
+                                state.dense_infos[key][state.dense_save_index - 1]
+                                for key in state.dense_infos.keys()
+                            ],
+                        )
+                    ),
```

Comment:
So `dense_infos` needn't be a dictionary; it can be any PyTree. If you need to index into it then you should do
```python
jtu.tree_map(lambda x: x[state.dense_save_index - 1], state.dense_infos)
```